torch/csrc/jit/ir.cpp

```diff
@@ -894,6 +894,171 @@ Node* Node::insertAfter(Node * n) {
+
+  // 1. Move from `this` toward movePoint, building up the working set of
+  // dependencies
+  std::list<Node*> workingSet;
+  workingSet.push_back(this);
+
+  int direction;
+  if (this->isAfter(movePoint)) {
+    direction = kPrevDirection;
+  } else {
+    direction = kNextDirection;
+  }
+
+  auto curNode = this->next_in_graph[direction];
+  // Move forward one node at a time
+  while (curNode != movePoint) {
+    if (curNode->isDependent(workingSet)) {
+      // If we can't move past this node, add it to the working set
+      workingSet.push_back(curNode);
+    }
+    curNode = curNode->next_in_graph[direction];
+  }
+
+  // 2. Decide whether we can move it all to `movePoint`.
+
+  // Say we are moving directly before movePoint and `this` starts before
+  // movePoint in the graph. The move looks like
+  //
+  //  `this`              `this`           |
+  //  <dependencies>  ->  `movePoint`      | `this` and deps are split
+  //  `movePoint`         <dependencies>   |
+  //
+  // Contrast with the case where `this` starts AFTER movePoint:
+  //
+  //  `movePoint`         <dependencies>   |
+  //  <dependencies>  ->  `this`           | `this` and deps are together
+  //  `this`              `movePoint`      |
+  //
+  // In the first case, we need to split `this` off from its dependencies, so we
+  // can move the dependencies below `movePoint` and keep `this` above.
+  const bool splitThisAndDeps =
+      (moveSide == MoveSide::BEFORE && this->isBefore(movePoint)) ||
+      (moveSide == MoveSide::AFTER && this->isAfter(movePoint));
+
+  if (splitThisAndDeps) {
+    // remove `this` from dependencies to be moved past `movePoint`
+    workingSet.pop_front();
+  }
+
+  // Check if we can move the working set past the move point
+  if (movePoint->isDependent(workingSet)) {
+    // if we can't, then there are intermediate dependencies between the
+    // `this` and `movePoint`, so we can't do the move
+    return false;
+  }
+
+  // 3. Execute the move
+  JIT_ASSERT(curNode == movePoint);
+  if (splitThisAndDeps) {
+    // Move `this`
+    this->move(movePoint, moveSide);
+
+    // Then move all of its dependencies on the other side of `movePoint`
+    const auto reversed =
+        moveSide == MoveSide::BEFORE ? MoveSide::AFTER : MoveSide::BEFORE;
+    for (auto toMove : workingSet) {
+      toMove->move(curNode, reversed);
+      curNode = toMove;
+    }
+  } else {
+    // Just append/prepend everything to `movePoint`
+    for (auto toMove : workingSet) {
+      toMove->move(curNode, moveSide);
+      curNode = toMove;
+    }
+  }
+  return true;
+}
+
+// Helper function so we can generalize `tryMove`
+void Node::move(Node* movePoint, MoveSide moveSide) {
+  switch (moveSide) {
+    case MoveSide::BEFORE:
+      this->moveBefore(movePoint);
+      break;
+    case MoveSide::AFTER:
+      this->moveAfter(movePoint);
+      break;
+  }
+}
+
+// Does `this` output any values used by `nodes`?
+template <typename T>
```

Comment:
originally it was because of recursion (wanted to work on std::list and graph_node_list when looking at inner blocks) but since the impl changed we don't need anymore